File: tests/test_work_graph_commit_paths.py
from __future__ import annotations

import subprocess
from pathlib import Path

import pytest

from entroly.work_graph_repo import RepositoryDiscoveryError, discover_repository_observation


def _git(repo: Path, *args: str) -> str:
    result = subprocess.run(
        ["git", "-C", str(repo), *args],
        check=True,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return result.stdout.strip()


def _repo(tmp_path: Path) -> Path:
    repo = tmp_path / "repo"
    repo.mkdir()
    _git(repo, "init", "-b", "main")
    _git(repo, "config", "user.email", "test@example.com")
    _git(repo, "config", "user.name", "Test")
    (repo / "base.txt").write_text("base\n", encoding="utf-8")
    _git(repo, "add", "base.txt")
    _git(repo, "commit", "-m", "base")
    _git(repo, "checkout", "-b", "feature/interrupted")
    return repo


def test_clean_ahead_branch_recovers_complete_per_commit_paths(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    (repo / "base.txt").write_text("changed\n", encoding="utf-8")
    (repo / "src").mkdir()
    (repo / "src" / "new.py").write_text("VALUE = 1\n", encoding="utf-8")
    _git(repo, "add", "base.txt", "src/new.py")
    _git(repo, "commit", "-m", "feature one")

    (repo / "docs.md").write_text("docs\n", encoding="utf-8")
    _git(repo, "add", "docs.md")
    _git(repo, "commit", "-m", "feature two")

    observation = discover_repository_observation(
        repo,
        include_checkpoint=False,
        observed_at_ms=1,
    )
    assert observation["changes"] == []
    assert observation["branch"]["ahead_by"] == 2
    assert len(observation["commits"]) == 2

    by_subject = {commit["subject"]: commit for commit in observation["commits"]}
    assert by_subject["feature one"]["changed_paths"] == ["base.txt", "src/new.py"]
    assert by_subject["feature two"]["changed_paths"] == ["docs.md"]


def test_commit_path_expansion_fails_closed_instead_of_returning_partial_history(
    tmp_path: Path,
) -> None:
    repo = _repo(tmp_path)
    bulk = repo / "bulk"
    bulk.mkdir()
    for index in range(513):
        (bulk / f"f{index:03d}.txt").write_text("x\n", encoding="utf-8")
    _git(repo, "add", "bulk")
    _git(repo, "commit", "-m", "large commit")

    with pytest.raises(RepositoryDiscoveryError, match="commit-path observation"):
        discover_repository_observation(
            repo,
            include_checkpoint=False,
            observed_at_ms=2,
        )

File: entroly-wasm/test_work_graph_commit_paths.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');
const {
  RepositoryDiscoveryError,
  discoverRepositoryObservation,
} = require('./js/work_graph_repo');

function assert(condition, message) {
  if (!condition) throw new Error(message || 'assertion failed');
}

function git(repo, ...args) {
  return execFileSync('git', ['-C', repo, ...args], {
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'pipe'],
  }).trim();
}

function makeRepo(root) {
  const repo = path.join(root, 'repo');
  fs.mkdirSync(repo);
  git(repo, 'init', '-b', 'main');
  git(repo, 'config', 'user.email', 'test@example.com');
  git(repo, 'config', 'user.name', 'Test');
  fs.writeFileSync(path.join(repo, 'base.txt'), 'base\n');
  git(repo, 'add', 'base.txt');
  git(repo, 'commit', '-m', 'base');
  git(repo, 'checkout', '-b', 'feature/interrupted');
  return repo;
}

const root = fs.mkdtempSync(path.join(os.tmpdir(), 'entroly-work-commit-paths-'));
try {
  const repo = makeRepo(root);
  fs.writeFileSync(path.join(repo, 'base.txt'), 'changed\n');
  fs.mkdirSync(path.join(repo, 'src'));
  fs.writeFileSync(path.join(repo, 'src', 'new.js'), 'module.exports = 1;\n');
  git(repo, 'add', 'base.txt', 'src/new.js');
  git(repo, 'commit', '-m', 'feature one');
  fs.writeFileSync(path.join(repo, 'docs.md'), 'docs\n');
  git(repo, 'add', 'docs.md');
  git(repo, 'commit', '-m', 'feature two');

  const observation = discoverRepositoryObservation(repo, {
    includeCheckpoint: false,
    observedAtMs: 1,
  });
  assert(observation.changes.length === 0, 'clean feature branch unexpectedly dirty');
  assert(observation.branch.ahead_by === 2, 'ahead count drift');
  assert(observation.commits.length === 2, 'commit count drift');
  const bySubject = new Map(observation.commits.map(commit => [commit.subject, commit]));
  assert(
    JSON.stringify(bySubject.get('feature one').changed_paths) === JSON.stringify(['base.txt', 'src/new.js']),
    'first commit changed paths are incomplete',
  );
  assert(
    JSON.stringify(bySubject.get('feature two').changed_paths) === JSON.stringify(['docs.md']),
    'second commit changed paths are incomplete',
  );

  const hugeRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'entroly-work-commit-paths-huge-'));
  try {
    const hugeRepo = makeRepo(hugeRoot);
    fs.mkdirSync(path.join(hugeRepo, 'bulk'));
    for (let index = 0; index < 513; index += 1) {
      fs.writeFileSync(path.join(hugeRepo, 'bulk', `f${String(index).padStart(3, '0')}.txt`), 'x\n');
    }
    git(hugeRepo, 'add', 'bulk');
    git(hugeRepo, 'commit', '-m', 'large commit');
    let rejected = false;
    try {
      discoverRepositoryObservation(hugeRepo, { includeCheckpoint: false, observedAtMs: 2 });
    } catch (error) {
      rejected = error instanceof RepositoryDiscoveryError && /commit-path observation/.test(error.message);
    }
    assert(rejected, 'oversized commit-path expansion returned partial history');
  } finally {
    fs.rmSync(hugeRoot, { recursive: true, force: true });
  }

  console.log('Work Graph npm committed-path recovery contract: PASS');
} finally {
  fs.rmSync(root, { recursive: true, force: true });
}
